### PR TITLE
Add message producer feature

### PR DESCRIPTION
**Pull request by sonnysideup**
Primitive changes that add AMQP publishing of `ContainerImageBuildStatus` state changes.

### CODE REVIEW

File: api/v1alpha1/containerimagebuild_types.go [hunk 1]
```diff
@@ -92,13 +93,24 @@ type ContainerImageBuildSpec struct {
 
 // ContainerImageBuildStatus defines the observed state of ContainerImageBuild
 type ContainerImageBuildStatus struct {
+	PreviousState    BuildState   `json:"-"` // NOTE: should we persist this value?
```
Comment (Secretions) [inline api/v1alpha1/containerimagebuild_types.go:96]: Curious how this is consumed.

File: controllers/containerimagebuild_controller.go [hunk 2]
```diff
@@ -49,13 +52,13 @@ func (r *ContainerImageBuildReconciler) Reconcile(req ctrl.Request) (ctrl.Result
 	}
 
 	// ignore resources that have been processed on start
-	if len(build.Status.State) != 0 {
+	if build.Status.State != "" {
```
Comment (Secretions) [inline controllers/containerimagebuild_controller.go:55]: Not familiar enough with the codebase to say this suggestion definitely makes sense, but wonder if it'd be better if these had an initial state ("Uninitialized" or something) so we can compare it to something concrete.

File: pkg/message/options.go [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+	}
+	if !ok {
+		return fmt.Errorf("broker %q is invalid (supported brokers: %v)", name, SupportedBrokers)
+	}
+
+	switch opts.Broker {
+	case "amqp":
+		if opts.AmqpURI == "" || opts.AmqpQueue == "" {
+			return fmt.Errorf("broker %q requires a uri and queue name", "amqp")
+		}
```
Comment (zs-ddl) [inline pkg/message/options.go:29]:
```suggestion
		}
	default:
	     return fmt.Errorf("broker %q is invalid (supported brokers: %v)", name, SupportedBrokers)
```

File: controllers/start.go [hunk 4]
```diff
@@ -50,12 +51,25 @@ func StartManager(metricsAddr string, enableLeaderElection bool) {
 		os.Exit(1)
 	}
 
+	var publisher message.Producer
+	if brokerOpts != nil {
+		setupLog.Info("Initializing message publisher")
+
+		publisher, err = message.NewProducer(brokerOpts)
+		if err != nil {
```
Comment (zs-ddl) [inline controllers/start.go:59]:
```suggestion
		if publisher, err = message.NewProducer(brokerOpts); err != nil {
```

File: api/v1alpha1/containerimagebuild_types.go (hunk 1 above)
Comment (sonnysideup) [reply api/v1alpha1/containerimagebuild_types.go:96]: This just seemed like a good piece of information to broadcast (i.e. before/after). I still need to write the consumer bits but I thought perhaps I will need to take separate actions if move from A -> B vs A -> C. I'm probably violating YAGNI.

File: controllers/containerimagebuild_controller.go (hunk 2 above)
Comment (sonnysideup) [reply controllers/containerimagebuild_controller.go:55]: Ideally, I'd like to leverage kubebuilder comment annotations to assign a default value (then we could do something like that) but I haven't been able to get it to work. I'm definitely going to spend some more time on this once the _first flush_ is out.

**Pull request merged**